Student dashboard/ active & inactive classes

eadams17 opened the pull request:
## WHAT
add two different sections: active and inactive classrooms on the student dashboard (rather than just one big classes section)

## WHY
we wanted to make it less confusing for students to know which are active classrooms

## HOW
use two different queries when returning active and inactive classrooms, update UI and add caching logic where relevant

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)
https://www.notion.so/quill/Student-active-inactive-classes-feature-292d42e6f941805ebdd2f02053879b4e?source=copy_link

### What have you done to QA this feature?
(Provide enough detail that a reviewer could assess whether additional QA should be done. For larger projects, additionally use the Engineer Feature Testing Notion template. Review Guidelines if needed: https://www.notion.so/quill/Github-PR-QA-Guidelines-49e99fc965654ceeb8c6249bd9d181d7)
tested on staging with various users to ensure that active/inactive classrooms are sorted as expected

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | yes
Have you deployed to Staging? | yes
Self-Review: Have you done an initial self-review of the code below on Github? | yes

eadams17 committed:
add active and inactive classes for students

eadams17 committed:
add/update specs

eadams17 committed:
fix tests

eadams17 committed:
fix spec

eadams17 committed:
tweak spec

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Student/components/selectAClassroom.tsx:
```diff
@@ -39,10 +43,38 @@ const SelectAClassroom: React.FC<SelectAClassroomProps> = ({
+    };
+    return (
+      <button
+        aria-label=""
```
Oof -- I see that this was already in the code, but we should remove this.

emilia-friedberg commented on services/QuillLMS/app/models/activity_session.rb:
```diff
@@ -587,4 +588,8 @@ def save_user_pack_sequence_items
+
+  private def invalidate_student_classrooms_cache
+    Rails.cache.delete("student_classrooms:#{user_id}") if user_id
+  end
```
Should this actually be invalidating the cache for every student associated with the classroom that this activity session is associated with?

emilia-friedberg commented on services/QuillLMS/app/models/classroom_unit.rb:
```diff
@@ -169,4 +170,11 @@ def save_user_pack_sequence_items
+    # Invalidate cache for all students assigned to this classroom unit
+    assigned_student_ids.each do |student_id|
+      Rails.cache.delete("student_classrooms:#{student_id}")
+    end
```
Like in activity session -- I think this should actually be every student in the class, not just ones that have been assigned?

emilia-friedberg reviewed (commented):
I'm curious about how much these queries are actually slowing down the load for this page. It might be worth messing with the queries themselves to see if we can optimize performance so that we don't need the caching layer at all. For instance, it might be faster to gather all the classrooms, along with the dates of last assignment and last session completion, and then just group them either on the front- or backend by activity status, rather than running two separate queries. I also think that would make it easier to rewrite these as active record queries, which might be preferable for maintenance. If the queries are really slow, this approach makes sense to me, but it just seems like we're going to end up invalidating the caches a lot anyway so maybe it would be better to just not have to maintain that logic.

eadams17 committed:
remove caching and change query to use active record

eadams17 commented:
> I'm curious about how much these queries are actually slowing down the load for this page. It might be worth messing with the queries themselves to see if we can optimize performance so that we don't need the caching layer at all. For instance, it might be faster to gather all the classrooms, along with the dates of last assignment and last session completion, and then just group them either on the front- or backend by activity status, rather than running two separate queries. I also think that would make it easier to rewrite these as active record queries, which might be preferable for maintenance. If the queries are really slow, this approach makes sense to me, but it just seems like we're going to end up invalidating the caches a lot anyway so maybe it would be better to just not have to maintain that logic.

I think you're right-- the caching is overkill. I added it because of [this note](https://www.notion.so/quill/Student-active-inactive-classes-feature-292d42e6f941805ebdd2f02053879b4e?source=copy_link#292d42e6f941806c83e5c97d4cba47d3) in the spec, but when I checked a few users with high classroom and activity counts, it didn't actually take long on initial load. I removed the caching logic and updated the queries to use ActiveRecord

eadams17 committed:
remove prop

emilia-friedberg reviewed (approved):
LGTM! Thanks for doing that rewrite -- I think this'll be easier to maintain.

emilia-friedberg commented on services/QuillLMS/app/controllers/profiles_controller.rb:
```diff
@@ -167,6 +170,106 @@ def staff
+        'last_assignments.last_assignment_date > ? OR ' \
+        'last_completions.last_completion_date > ?',
+        one_year_ago,
+        six_months_ago,
+        six_months_ago
+      )
+      .select('classrooms.name, teachers.name AS teacher, classrooms.id, sc_dates.join_date')
+      .distinct
+      .order('sc_dates.join_date DESC')
+      .map { |c| { 'name' => c.name, 'teacher' => c.teacher, 'id' => c.id } }
+  end
+
+  protected def inactive_students_classrooms
+    one_year_ago = 365.days.ago
+    six_months_ago = 180.days.ago
+
+    # Get student's classroom IDs for scoping
+    student_classroom_ids = current_user.students_classrooms.pluck(:classroom_id)
+
+    return [] if student_classroom_ids.empty?
+
+    # Build scoped subquery for last assignment date
+    last_assignment_subquery = ClassroomUnit
+      .where(classroom_id: student_classroom_ids, visible: true)
+      .group(:classroom_id)
+      .select('classroom_id, MAX(created_at) as last_assignment_date')
+
+    # Build scoped subquery for last completion date
+    last_completion_subquery = ActivitySession
+      .joins(:classroom_unit)
+      .where(classroom_units: { classroom_id: student_classroom_ids, visible: true })
+      .where(activity_sessions: { visible: true, state: 'finished' })
+      .where.not(activity_sessions: { completed_at: nil })
+      .group('classroom_units.classroom_id')
+      .select('classroom_units.classroom_id, MAX(activity_sessions.completed_at) as last_completion_date')
+
+    # Main query - get student_classroom join date in subquery for proper ordering
+    join_date_subquery = StudentsClassrooms
+      .where(student_id: current_user.id)
+      .select('classroom_id, created_at as join_date')
```
One additional note is that I think these are shared across methods, so you could break them out into their own functions and just call them in both -- you could probably just ask Claude to consolidate shared code here (there's more in the main query itself that's identical). Then we don't need to update code in two separate places if these queries require changes later.

eadams17 committed:
combined shared query logic

anathomical reviewed (approved):
I'm a little concerned about the complexity of the multiple queries being chained together, but nothing here seems like a huge issue.  I took a stab at a version of the query using `HAVING` clauses that is more readable, but running with a couple of test users shows that the version of this that I have is not 1-to-1 in results.  I don't think that it's worth spending a ton of time trying to massage my updated query to get exactly what we want, but I'm going to leave it here for posterity in case anyone ever comes back to refactor this code.

```ruby
def base_query
  Classroom
    .select('classrooms.id, classrooms.name, users.name AS teacher, students_classrooms.created_at AS join_date')
    .joins(classroom_units: :activity_sessions)
    .joins(classrooms_teachers: :user)
    .joins(:students_classrooms)
    .where(classroom_units: { visible: true, activity_sessions: { visible: true } })
    .where(classrooms_teachers: { role: 'owner' })
    .where(students_classrooms: { student_id: current_user.id })
    .group('classrooms.id, classrooms.name, users.name, students_classrooms.created_at')
    .order('students_classrooms.created_at DESC')
end

def active_students_classrooms
  having_clause = <<-SQL
    MAX(classrooms.created_at) > ? OR
    MAX(classroom_units.created_at) > ? OR
    MAX(activity_sessions.completed_at) > ?
  SQL

  base_query.having(having_clause, one_year_ago, six_months_ago, six_months_ago)
end

def inactive_students_classrooms
  having_clause = <<-SQL
    MAX(classroom_units.created_at) IS NULL OR
    (MAX(classroom_units.created_at) <= ? AND
    (MAX(activity_sessions.completed_at) IS NULL OR MAX(activity_sessions.completed_at) <= ?))
  SQL

  base_query.having(having_clause, six_months_ago, six_months_ago)
end
```

eadams17 committed:
remove unused code

eadams17 committed:
Merge branch 'develop' of https://github.com/empirical-org/Empirical-Core into student-active-inactive-classes
